spec/requests/omniauth_callbacks_spec.rb

```diff
@@ -0,0 +1,382 @@
+          user = create(:user, email: oauth_email, third_party_avatar: 'http://example.com/images/some-image.jpeg')
+          avatar_image_url = 'http://example.com/images/different_profile.jpeg'
+          set_avatar_env(avatar_image_url)
+
+          get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+          expect(user.reload.third_party_avatar).to eq(avatar_image_url)
+        end
+
+        it 'uploads avatar when third_party_avatar is nil' do
+          user = create(:user, email: oauth_email, third_party_avatar: nil)
+          avatar_image_url = 'http://example.com/images/different_profile.jpeg'
+          set_avatar_env(avatar_image_url)
+
+          get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+          expect(user.reload.third_party_avatar).to eq(avatar_image_url)
+        end
+
+        it 'does not upload third party avatar if current avatar is the same' do
+          avatar_image_url = 'http://example.com/images/profile.jpeg'
+          user = create(:user, email: oauth_email, third_party_avatar: avatar_image_url)
+          set_avatar_env(avatar_image_url)
+
+          expect {
+            get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+          }.not_to change {
+            user.reload.third_party_avatar
+          }
+        end
+      end
+
+      context 'when third party avatar is nil' do
+        it 'does not set third_party_avatar' do
+          set_avatar_env(nil)
+
+          get omniauth_login_omniauth_callbacks_path, params: { provider: 'google' }
+
+          user = User.find_by(email: oauth_email)
+          expect(user.third_party_avatar).to be_nil
+        end
+      end
+    end
+  end
+
+  describe 'GET #facebook' do
+    before do
+      set_devise_mapping_env
+      set_omniauth_auth_env(provider: 'facebook')
+    end
+
+    context 'when facebook email doesnt exist in the system' do
+      it 'creates user with info in facebook' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'facebook' }
+
+        user = User.find_by(email: oauth_email)
+        expect(user).to be_present
+      end
+
+      it 'signs in and redirect to root' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'facebook' }
+
+        expect(response).to redirect_to root_path
+      end
+
+      it 'enables oauth on user' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'facebook' }
+
+        user = User.find_by(email: oauth_email)
+        expect(user.oauth_enabled?).to eq true
+      end
+    end
+
+    context 'when facebook email already exist in the system' do
+      let!(:user) { create(:user, email: oauth_email, token: nil) }
+
+      it 'updates the user with facebook credentials' do
+        get omniauth_login_omniauth_callbacks_path, params: { provider: 'facebook' }
+
+        expect(user.reload.token).to eq 'abcdefg12345'
```

Comment:
I feel like `'abcdefg12345'` is a magic value code smell. It'd be better if there was a variable that reference this somewhere.